### PR TITLE
=ddata Garbage collect valueDeltas tombstones at merge/mergeDelta (#22974)

### PR DESCRIPTION
**Pull request by gosubpl**
* tombstones created for withValueDeltas ORMultiMap are garbage collected at each merge/mergeDelta operation
* this avoids tombstones resulting in permanent resource leak

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #22974

**Comment by gosubpl** [comment]
Marked *wip* as I have put it on repeat testing and also want test on some sample projects to avoid regression.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala [hunk 1]
```diff
@@ -69,9 +69,12 @@ final class ORMultiMap[A, B] private[akka] (
 
   override def merge(that: T): T =
     if (withValueDeltas == that.withValueDeltas) {
-      if (withValueDeltas)
-        new ORMultiMap(underlying.mergeRetainingDeletedValues(that.underlying), withValueDeltas)
-      else
+      if (withValueDeltas) {
+        val newUnderlying = underlying.mergeRetainingDeletedValues(that.underlying)
+        // Garbage collect the tombstones we no longer need, i.e. those that have Set() as a value.
+        val newValues = newUnderlying.values.filterNot(elem ⇒ (!newUnderlying.keys.contains(elem._1)) && elem._2.isEmpty)
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala]: `filterNot { case (key, value) =>`

**Review by patriknw — COMMENTED** [review]
looking good, update documentation also

I'll probably take another look later

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala (hunk 1 above)
Comment (gosubpl) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala]: Thanks. Tried to figure it out, but forgot, that in Scala you need to use curly braces (`{`) to enable this. Why does this language need to have so irregular grammar...

**Comment by gosubpl** [comment]
@patriknw initial comments reflected. Thanks!

**Comment by gosubpl** [comment]
Tested on repeat with various tests, no longer wip.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala (hunk 1 above)
Comment (ktoso) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala]: Martin would disagree :D But yeah, the dichotomy between partial and functions is no fun and will hopefully be addressed at some point sooner rather than later

Comment (gosubpl) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala]: I am probably too harsh here, but my first 'serious' programming language (after BASIC and 6502 Assembly) was Pascal, and I am missing the consistency and elegance ever since :)

**Review by raboof — APPROVED** [review]
Looks good. The clear invariants of CRDT's seem to make it a great candidate for some property-based tests to fuzz out corner cases, I don't think we have any of these, right? Would make a great addition (but for another PR)

**Comment by gosubpl** [comment]
@raboof what would you recommend to do property-based / fuzz testing? Any machinery already in Akka testing scaffolding?
I am now learning https://github.com/epfl-lara/stainless . What would you say to adding a Stainless Spec (I am still not sure how would that work :) ) ?

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
thanks @gosubpl

**Comment by raboof** [comment]
@gosubpl I think scalacheck is the common tool for property-based testing - we use it in a couple of places (e.g. `ByteStringSpec`).

Stainless sounds super cool - if we could use that to prove some properties that'd be amazing. My (outdated) experience with provers is that problems quickly become too big to reasonably cover, but DData might just be the 'right-sized' problem. Would be a fun experiment!
